File: Cpp/LeetCode3.LongestSubstringWithoutRepeatingCharacters_SlidingWindow.cpp
// using Sliding Window with 2 pointers
// Time Complexity: O(N)
// Space Complexity: O(N)

#include <string>
#include <unordered_map>

class Solution {
public:
    int lengthOfLongestSubstring(std::string s) {
        int left = 0; // left index of window
        int right = 0; // right index of window
        int matched = 0; // to check how many types characters already fullfilled the target
        std::unordered_map<char, int> window; // to store amount of each target character found in window
        int longest = 0;
        int sStringSize = s.size();
        
        while (right < sStringSize) {
            char c = s[right];
            right++;
            window[c]++;
            
            // shrink when the window contained the character more than once
            while (window[c]>1) {
                char d = s[left];
                left++;
                window[d]--;
            }
            
            if (right - left > longest) {
                longest = right - left ;
            }
            
        }
        return longest;

    }
};

File: Python/LeetCode3.LongestSubstringWithoutRepeatingCharacters_SlidingWindow.py
# Time Complexity: O(N)
# Space Complexity: O(N)

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left, right, longest = 0, 0, 0
        char_count = {}

        while right < len(s):
            a = s[right]
            right += 1

            char_count[a] = char_count.get(a, 0) + 1

            while char_count[a] > 1:
                char_count[s[left]] -= 1
                left += 1

            longest = max(longest, right - left)

        return longest
